File: samples/csharp_equiv/combat.cs
using System;

class Entity
{
    public int Hp { get; set; }
    public int Atk { get; }
    public int Def { get; }

    public Entity(int hp, int atk, int def)
    {
        Hp = hp;
        Atk = atk;
        Def = def;
    }
}

class Program
{
    static Entity MakeEntity(string name, int hp, int atk, int def)
    {
        return new Entity(hp, atk, def);
    }

    static Entity ApplyDmg(Entity ent, int amt)
    {
        int oh = ent.Hp;
        int raw = amt - ent.Def;
        int dmg = raw < 0 ? 0 : raw;
        ent.Hp = Math.Max(0, oh - dmg);
        return ent;
    }

    static bool IsAlive(Entity ent) => ent.Hp > 0;

    static void Main()
    {
        var hero = MakeEntity("Hero", 100, 25, 10);
        var goblin = MakeEntity("Goblin", 30, 15, 5);

        goblin = ApplyDmg(goblin, hero.Atk);
        Console.WriteLine($"Goblin HP after hero attack: {goblin.Hp}");
        Console.WriteLine($"Goblin alive? {IsAlive(goblin)}");
    }
}

File: samples/python_equiv/combat.py
class Entity:
    def __init__(self, hp, atk, defense):
        self.hp = hp
        self.atk = atk
        self.defense = defense

def make_entity(name, hp, atk, defense):
    return Entity(hp, atk, defense)

def apply_dmg(ent, amt):
    assert amt >= 0, "precondition: amt >= 0"
    oh = ent.hp
    raw = amt - ent.defense
    dmg = 0 if raw < 0 else raw
    ent.hp = max(0, oh - dmg)
    assert ent.hp <= oh, "postcondition: hp <= old hp"
    return ent

def is_alive(ent):
    return ent.hp > 0

hero = make_entity("Hero", 100, 25, 10)
goblin = make_entity("Goblin", 30, 15, 5)

goblin = apply_dmg(goblin, hero.atk)
print("Goblin HP after hero attack:", goblin.hp)
print("Goblin alive?", is_alive(goblin))
